clippy_lints/src/formatting.rs

```diff
@@ -133,6 +157,49 @@ fn check_assign(cx: &EarlyContext<'_>, expr: &Expr) {
     }
 }
 
+/// Implementation of the `SUSPICIOUS_UNARY_OP_FORMATTING` lint.
+fn check_unop(cx: &EarlyContext<'_>, expr: &Expr) {
+    if let ExprKind::Binary(ref binop, ref lhs, ref rhs) = expr.kind {
+        if !differing_macro_contexts(lhs.span, rhs.span) && !lhs.span.from_expansion() {
+            // span between BinOp LHS and RHS
+            let binop_span = lhs.span.between(rhs.span);
+            // if RHS is a UnOp
+            if let ExprKind::Unary(op, ref un_rhs) = rhs.kind {
+                // from UnOp operator to UnOp operand
+                let unop_operand_span = rhs.span.until(un_rhs.span);
+                if let (Some(binop_snippet), Some(unop_operand_snippet)) =
+                    (snippet_opt(cx, binop_span), snippet_opt(cx, unop_operand_span))
+                {
+                    let binop_str = BinOpKind::to_string(&binop.node);
+                    // no space after BinOp operator and space after UnOp operator
+                    if binop_snippet.ends_with(binop_str) && unop_operand_snippet.ends_with(' ') {
+                        let unop_str = UnOp::to_string(op);
+                        let eqop_span = lhs.span.between(un_rhs.span);
+                        span_note_and_lint(
+                            cx,
+                            SUSPICIOUS_UNARY_OP_FORMATTING,
+                            eqop_span,
+                            &format!(
+                                "by not having a space between `{binop}` and `{unop}` it looks like \
+                                 `{binop}{unop}` is a single operator",
+                                binop = binop_str,
+                                unop = unop_str
+                            ),
+                            eqop_span,
+                            &format!(
+                                "to remove this lint, put a space between `{binop}` and `{unop}` \
+                                 or remove the space after `{binop}`",
```

Comment:
```suggestion
                                 and remove the space after `{binop}`",
```

Reply:
This suggests that we also lint for something like `x > - 1` which we currently don't. Should we?
By the way I realized this should be ```remove the space after `{unop}` ``` rather than `{binop}`. Ideally there should be a space after `{binop}`.

Reply:
> This suggests that we also lint for something like x > - 1 which we currently don't. Should we?

No we shouldn't lint that IMO, since isn't as confusing as `&&!` or similar and just a fmt thing.

But the lint triggers only on formatting where the unop is directly after a binop **and** with (at least?) one space between the unop and the operator, IIUC. So we could directly suggest to do both?

Reply:
I see your point, as the text is now a suggestion and not a way to remove the lint as it was before (for which either removing a space or adding one would suffice).
Changed the text.